examples/provider_upgrade_test.go

```diff
@@ -0,0 +1,322 @@
+
+	previewCounter := 0
+	var pt *integration.ProgramTester
+	test := opts.With(integration.ProgramTestOptions{
+		Env: env,
+		PrePulumiCommand: func(verb string) (func(err error) error, error) {
+			if verb != "preview" {
+				return nil, nil
+			}
+			previewCounter++
+			if previewCounter != 2 {
+				return nil, nil
+			}
+			t.Logf("Importing pre-recorded stateFile")
+			importState(t, pt, info.stateFile)
+			return nil, nil
+		},
+		SkipRefresh:      true,
+		SkipExportImport: true,
+
+		// TODO this test is not quite right. There's two designs for an upgrade test, a
+		// preview-only one design and an full update design. This is currently neither.
+		//
+		// A preview-only design would import V5 state and run pulumi preview on V6 to
+		// assert empty preview. To make this code behave like this we need to run
+		// SkipUpdate: true but this currently messes up with stack name renaming pass that
+		//
+		// SkipUpdate:       true,
+		//
+		// A full update design would provision actual cloud infra on V5 and do a full
+		// pulumi-up on V6. This can be useful if for example update plans are permitted
+		// (only replaces are rejected) but we need to verify that the actual update plans
+		// execute successfully or if the provider update does not touch the infra but
+		// updates the state file and we need to test that.
+		//
+		// What this code does: provision infra on V6, then import V5 state, then do pulumi
+		// up. This mostly works as a preview test, but does unnecessary "up" work.
+	})
+	pt = integration.ProgramTestManualLifeCycle(t, &test)
+	err := pt.TestLifeCycleInitAndDestroy()
+	require.NoError(t, err)
+}
+
+type providerUpgradeInfo struct {
+	testCaseDir             string
+	recordingDir            string
+	baselineProviderVersion string
+	grpcFile                string
+	stateFile               string
+	opts                    integration.ProgramTestOptions
+}
+
+func newProviderUpgradeInfo(t *testing.T) providerUpgradeInfo {
+	info := providerUpgradeInfo{}
+	info.testCaseDir = filepath.Join("testdata", "resources")
+	pyaml := filepath.Join(info.testCaseDir, "Pulumi.yaml")
+	v := parseProviderVersion(t, pyaml)
+	info.baselineProviderVersion = v
+	info.recordingDir = filepath.Join("testdata", "recorded", v, "resources")
+	var err error
+	info.grpcFile, err = filepath.Abs(filepath.Join(info.recordingDir, "grpc.json"))
+	require.NoError(t, err)
+	info.stateFile, err = filepath.Abs(filepath.Join(info.recordingDir, "state.json"))
+	require.NoError(t, err)
+	info.opts = integration.ProgramTestOptions{
+		Dir: info.testCaseDir,
+	}
+	return info
+}
+
+func importState(t *testing.T, pt *integration.ProgramTester, stateFile string) {
+	tempDir := t.TempDir()
+	newStateFile := filepath.Join(tempDir, "new-state.json")
+	pt.RunPulumiCommand("stack", "export", "--file", newStateFile)
+	stackName := parseStackName(t, readFile(t, newStateFile))
+	fixedState := withUpdatedStackName(t, stackName, readFile(t, stateFile))
```

Comment:
Well no, I think in fact this would invalidate the test.. We want to read the pre-recorded stateFile (from stateFile variable). We're only reading the newStateFile as a workaround to determine currently allocated stack's name. I can factor it out into a function to make more obvious.